agents/tests/test_dns_agent.py

```diff
@@ -0,0 +1,163 @@
+
+        assert set(result.keys()) == required_keys
+
+
+def test_status_to_dict_with_missing_fields(dns_agent):
+    """
+    Ensure status_to_dict() handles missing or None fields gracefully.
+    """
+    with patch.object(dns_agent, 'collect_server_metadata'):
+        dns_agent.os_type = None
+        dns_agent.hostname = None
+        dns_agent.ip = None
+        dns_agent.server_name = 'dns'
+        dns_agent.uptime = -1
+        dns_agent.timestamp = None
+        dns_agent.healthy = False
+
+        result = dns_agent.status_to_dict()
+
+        assert result['os'] is None
+        assert result['hostname'] is None
+        assert result['ip'] is None
+        assert result['uptime'] == -1
+        assert result['healthy'] is False
+
+
+@patch('subprocess.Popen')
+def test_is_dns_running_success(mock_popen, dns_agent):
+    """
+    Test that is_dns_running()
+    returns True when the dig command succeeds with output.
+    """
+    process_mock = MagicMock()
+    process_mock.communicate.return_value = (b'8.8.8.8\n', b'')
+    process_mock.returncode = 0
+    mock_popen.return_value = process_mock
+
+    assert dns_agent.is_dns_running() is True
+
+
+@patch('subprocess.Popen')
+def test_is_dns_running_no_output(mock_popen, dns_agent):
+    """
+    Test that is_dns_running() returns False when dig returns 0 but no output.
+    """
+    process_mock = MagicMock()
+    process_mock.communicate.return_value = (b'', b'')
+    process_mock.returncode = 0
+    mock_popen.return_value = process_mock
+
+    assert dns_agent.is_dns_running() is False
+
+
+@patch('subprocess.Popen')
+def test_is_dns_running_error_code(mock_popen, dns_agent):
+    """
+    Test that is_dns_running()
+    returns False when the dig command fails or returns an error.
+    """
+    process_mock = MagicMock()
+    process_mock.communicate.return_value = (b'', b'Some error')
+    process_mock.returncode = 1
+    mock_popen.return_value = process_mock
+
+    assert dns_agent.is_dns_running() is False
+
+
+@patch.object(DNSAgent, '_is_process_running', return_value=True)
+@patch.object(DNSAgent, '_is_port_open', return_value=True)
+@patch.object(DNSAgent, 'is_dns_running', return_value=True)
+def test_service_healthy_true(mock_dns, mock_port, mock_proc, dns_agent):
+    """
+    Test service_healthy()
+    returns True when all checks (process, port, DNS) pass.
+    """
+    assert dns_agent.service_healthy() is True
+
+
+@patch.object(DNSAgent, '_is_process_running', return_value=False)
+@patch.object(DNSAgent, '_is_port_open', return_value=True)
+@patch.object(DNSAgent, 'is_dns_running', return_value=True)
+def test_service_healthy_fails_due_to_process(
+    mock_dns,
+    mock_port,
+    mock_proc,
+    dns_agent,
+):
+    """
+    Test service_healthy()
+    returns False when the process check fails, even if others pass.
+    """
+    assert dns_agent.service_healthy() is False
+
+
+@patch('agents.dns.dns.maybe_log_message')
```

Comment:
Optional but it would just be better to actually log message to the tempfile - especially that you did create the temple in `dns_agent` fixture.